lib/utils/logger.ts

```diff
@@ -114,6 +123,26 @@ class Logger {
 	public logLivepush(msg: string) {
 		return this.streams.livepush.write(msg + eol);
 	}
+
+	/**
+	 * Log a warning for display later, ignore duplicates.
+	 */
+	public deferredWarn(msg: string) {
```

Comment:
Nice. Minor suggestion: instead of `deferredWarn(msg)` (whose name is similar to `logWarn()` and suggests adding `deferredInfo()`, `deferredError()`...), it could be `deferredLog(msg, level)`, where `level` is one of `info, warn, error...`, and a single array of `[msg, level]` pairs:

```
	public deferredLog(msg: string, level: string) {
		if (!this.deferredLogMessages.find(entry => entry[0] === msg)) {
			this.deferredLogMessages.push(msg, level);
		}
	}

	public outputDeferredMessages() {
		this.deferredLogMessages.forEach((msg, level) => {
			this.streams[level].write(msg + eol);
		});
	}
```

Reply:
Right, I'd considered this but decided to keep the the signature the same as the other functions (and indeed expected others to be added if needed ;).   Can change though, sure.